src/service/createUpdateIdentifierPersona.ts

```diff
@@ -0,0 +1,112 @@
+import NoModel from 'jscommons/dist/errors/NoModel';
+import * as promiseRetry from 'promise-retry';
+import Locked from '../errors/Locked';
+import CreateUpdateIdentifierPersonaOptions // tslint:disable-line:import-spacing
+  from '../serviceFactory/options/CreateUpdateIdentifierPersonaOptions';
+import CreateUpdateIdentifierPersonaResult // tslint:disable-line:import-spacing
+  from '../serviceFactory/results/CreateUpdateIdentifierPersonaResult';
+import Config from './Config';
+import service from './index';
+
+const create = (config: Config) =>
+  async ({organisation, ifi, personaName}: CreateUpdateIdentifierPersonaOptions):
+  Promise<CreateUpdateIdentifierPersonaResult> => {
+    const { identifier } = await config.repo.createIdentifier({
+      ifi,
+      locked: true,
+      organisation,
+    });
+
+    const { persona } = await service(config).createPersona({
+      name: personaName,
+      organisation,
+    });
+
+    await config.repo.setIdentifierPersona({
+      id: identifier.id,
+      locked: false,
+      organisation,
+      persona: persona.id,
+    });
+
+    return {
+      identifierId: identifier.id,
+      personaId: persona.id,
+    };
+  };
+
+const createUpdateIdentifierPersona = (config: Config) =>
+  async ({
+    organisation,
+    ifi,
+    personaName,
+  }: CreateUpdateIdentifierPersonaOptions): Promise<CreateUpdateIdentifierPersonaResult> => {
+
+    // find the ifi
+    try { // Identifier exists, so do nothing.
+      const { identifierId } = await service(config).getIdentifierByIfi({
+        ifi,
+        organisation,
+      });
+      const {identifier: foundIdentifier, locked} = await config.repo.getIdentifier({
+        id: identifierId,
+        organisation,
+      });
+
+      if ( locked === true ) {
+        // We are locked, wait for unlock
+        throw new Locked();
+      }
+
+      // Shouldn't happen, as persona should always be set if unlocked.
+      /* istanbul ignore next */
+      if (foundIdentifier.persona === undefined) {
+        throw new Error('Identifier should have a persona');
+      }
+
+      // What should happen if persona name is different ???
+      // currently it doesn't get updated
+
+      return {
+        identifierId,
+        personaId: foundIdentifier.persona,
+      };
+
+    } catch (err) {
+      if (err instanceof NoModel) {
+        return create(config)({
+          ifi,
+          organisation,
+          personaName,
+        });
+      }
+      throw err;
+    }
+
+  };
+
+const retryCreateUpdateIdentifierPersona = (config: Config) =>
```

Comment:
My gut feeling is that there's a better way to do this considering we'd never create identifiers without also ensuring a persona is on there, I need to have more of a think about it though right now. The fact that you're calling the service inside the service and the fact that there is stuff in the service here that is aware of the locking is triggering some alarm bells for me @andrewhickey @cbishopvelti.

Reply:
Fixed calling service within the service.

Reply:
I have no problem with the locking logic being in the service, it's agnostic to the repo, it's not exposed to the interface. Don't see the problem.

Reply:
Should it be agnostic to the repo though? Different repos might lock in different ways. It seems to me that perhaps the creation of the identifier and persona together, should be a repo function (perhaps createIdentifierWithPersona). That way if we switched to SQL it would be done in one transaction.